File: sandbox-runner/patterns/batch-iterator/typescript/activities.ts
import { TOTAL_RECORDS } from "./shared";

export async function fetchPage(offset: number, pageSize: number): Promise<number[]> {
  const end = Math.min(offset + pageSize, TOTAL_RECORDS);
  const page: number[] = [];
  for (let i = offset; i < end; i++) {
    page.push(i);
  }
  return page;
}

export async function processRecord(recordId: number): Promise<void> {
  // Simulate processing work.
  await new Promise((r) => setTimeout(r, 50));
}

File: sandbox-runner/patterns/batch-iterator/python/activities.py
import asyncio

from temporalio import activity

from shared import TOTAL_RECORDS


@activity.defn
async def fetch_page(offset: int, page_size: int) -> list[int]:
    end = min(offset + page_size, TOTAL_RECORDS)
    return list(range(offset, end))


@activity.defn
async def process_record(record_id: int) -> None:
    # Simulate processing work.
    await asyncio.sleep(0.05)
